src/CodegenShapeMember.php

```diff
@@ -0,0 +1,35 @@
+<?hh // strict
+/*
+ *  Copyright (c) 2015-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+ namespace Facebook\HackCodegen;
+
+ final class CodegenShapeMember {
+
+   public function __construct(
+     public string $name,
+     private mixed $type,
+     public bool $optional = false,
+   ) {
+     invariant(
+       is_string($type) || $type instanceof CodegenShapeFuture || $type instanceof CodegenShape,
```

Comment:
not sure if there is a better way to represent a union type like this

Reply:
- in this project: properties should be private, with public setter/getter methods 
- there's not a better approach for unions, but this should be a `CodegenType` - and `CodegenType` should be extended to support this future shape class
- we generally consider bool flags an antipattern as it makes callers unreadable unless they docuemnt the parameters or readers are familiar with the API; usual approaches are:
  - add an enum
  - remove the parameter, and require people to use a setter if a field is optional.

I'd probably go for the latter here.

Reply:
if this is `CodegenType` and we called `render()` on it, that would output:

```
type name = int
```

which doesn't seem like what we would want? i see how we would want to add support for `CodegenShape_FUTURE` within `CodegenType`, but not how we would use that here.

Reply:
ie. i'm not sure how with a constructor of:

```
final class CodegenShapeMember {
  public function __construct(
     private string $name,
     private CodegenType $type,
  ) {}
}
```
we could render `$type` within the shape.

Reply:
ah, yes, you're right. I forgot that `CodegenType` actually means a type declaration, not a type specifier :/

but CodegenType should also support these new shapes :)